tensornetwork/backends/jax/jax_backend_test.py

```diff
@@ -768,10 +795,121 @@ def mv(x, H):
+
+  thresh = {
+      np.complex64: 1E-3,
+      np.float32: 1E-3,
+      np.float64: 1E-4,
+      np.complex128: 1E-4
+  }
+  rtol = 1E-8
+  atol = 1E-8
   compare_eigvals_and_eigvecs(
-      np.stack(U, axis=1), eta, U_exact, eta_exact, thresh=1E-4, atol=1E-4)
+      np.stack(U, axis=1),
+      eta,
+      U_exact,
+      eta_exact,
+      rtol,
+      atol,
+      thresh=thresh[dtype])
+
+
+def get_ham_params(dtype, N, which):
+  if which == 'uniform':
+    hop = -jnp.ones(N - 1, dtype=dtype)
+    pot = jnp.ones(N, dtype=dtype)
+    if dtype in (np.complex128, np.complex64):
+      hop -= 1j * jnp.ones(N - 1, dtype)
+  elif which == 'rand':
+    hop = -jnp.array(np.random.rand(N - 1).astype(dtype)-0.5)
+    pot = jnp.array(np.random.rand(N).astype(dtype))-0.5
+    if dtype in (np.complex128, np.complex64):
+      hop -= 1j * jnp.array(np.random.rand(N - 1).astype(dtype)-0.5)
+  return pot, hop
+
+
+@pytest.mark.parametrize("dtype", [np.float64, np.complex128])
+@pytest.mark.parametrize("param_type", ['uniform', 'rand'])
+@pytest.mark.parametrize("N", [14])
+def test_eigsh_free_fermions(N, dtype, param_type):
+  """
+  Find the lowest eigenvalues and eigenvectors
+  of a 1d free-fermion Hamiltonian on N sites.
+  The dimension of the hermitian matrix is
+  (2**N, 2**N).
+  """
+  backend = jax_backend.JaxBackend(precision=jax.lax.Precision.HIGHEST)
+  np.random.seed(10)
+  pot, hop = get_ham_params(dtype, N, param_type)
+  P = jnp.diag(np.array([0, -1])).astype(dtype)
+  c = jnp.array([[0, 1], [0, 0]], dtype)
+  n = c.T @ c
+  eye = jnp.eye(2, dtype=dtype)
+  neye = jnp.kron(n, eye)
+  eyen = jnp.kron(eye, n)
+  ccT = jnp.kron(c @ P, c.T)
+  cTc = jnp.kron(c.T, c)
+
+  @jax.jit
+  def matvec(vec):
+    x = vec.reshape((4, 2**(N - 2)))
+    out = jnp.zeros(x.shape, x.dtype)
+    t1 = neye * pot[0] + eyen * pot[1] / 2
+    t2 = cTc * hop[0] - ccT * jnp.conj(hop[0])
+    out += jnp.einsum('ij,ki -> kj', x, t1 + t2)
+    x = x.reshape((2, 2**(N - 1))).transpose((1, 0)).reshape((4, 2**(N - 2)))
+    out = out.reshape((2, 2**(N - 1))).transpose((1, 0)).reshape(
+        (4, 2**(N - 2)))
+    for site in range(1, N - 2):
+      t1 = neye * pot[site] / 2 + eyen * pot[site + 1] / 2
+      t2 = cTc * hop[site] - ccT * jnp.conj(hop[site])
+      out += jnp.einsum('ij,ki -> kj', x, t1 + t2)
+      x = x.reshape((2, 2**(N - 1))).transpose((1, 0)).reshape((4, 2**(N - 2)))
+      out = out.reshape((2, 2**(N - 1))).transpose((1, 0)).reshape(
+          (4, 2**(N - 2)))
+    t1 = neye * pot[N - 2] / 2 + eyen * pot[N - 1]
+    t2 = cTc * hop[N - 2] - ccT * jnp.conj(hop[N - 2])
+    out += jnp.einsum('ij,ki -> kj', x, t1 + t2)
+    x = x.reshape((2, 2**(N - 1))).transpose((1, 0)).reshape((4, 2**(N - 2)))
+    out = out.reshape((2, 2**(N - 1))).transpose((1, 0)).reshape(
+        (4, 2**(N - 2)))
+
+    x = x.reshape((2, 2**(N - 1))).transpose((1, 0)).reshape(2**N)
+    out = out.reshape((2, 2**(N - 1))).transpose((1, 0)).reshape(2**N)
+    return out.ravel()
```

Comment:
What in the world is happening here lol.

Reply:
This is the matric vector product for the transverse field Ising Hamiltonian on open boundary conditions!

Reply:
I'll take your word for it.